datastream repopulation logic changes and removal of batch partially …

hexoscott opened the pull request:
…processed

partially processed removed in favour of simply sealing the WIP batch on a restart

hexoscott committed:
datastream repopulation logic changes and removal of batch partially processed

partially processed removed in favour of simply sealing the WIP batch on a restart

mandrigin reviewed (dismissed):
looks quite concise and nice

kstoykov commented on zk/stages/stage_sequence_execute.go:
```diff
@@ -79,29 +75,69 @@ func SpawnSequencingStage(
-
-	if err := utils.UpdateZkEVMBlockCfg(cfg.chainConfig, sdb.hermezDb, logPrefix); err != nil {
-		return err
+	if batchState.hasExecutorForThisBatch {
```
Can we make this in a function like "doCheckForBadBatch"?

hexoscott replied:
I'll take a look

kstoykov commented on zk/stages/stage_sequence_execute.go:
```diff
@@ -79,29 +75,69 @@ func SpawnSequencingStage(
-		// this occurs when sequencer was switched from syncer or sequencer datastream files were deleted
-		// and datastream was regenerated
-		if err = finalizeLastBatchInDatastreamIfNotFinalized(batchContext, batchState, executionAt); err != nil {
+	if !lastBatchSealed {
```
Can we just use "runBlocksLoops = false" in this case and let the rest of the code execute instead of creating a "special" case only for closing a batch. The goal is to eliminate the duplicate code.

hexoscott replied:
That sounds like a good plan!  Nice idea

hexoscott replied:
I made this change but it gets confusing, and for the code we drop we need to populate more in it's place.  Skipping the block loop causes the new batch to be sealed, but we want to seal the last batch instead.  I think it's more readable to have this self contained bit of code that is very explicit in what it's doing

kstoykov replied:
Ok

kstoykov commented on zk/stages/stage_sequence_execute_batch.go:
```diff
@@ -12,28 +12,7 @@ import (
-		}
-	}
-
+func prepareBatchCounters(batchContext *BatchContext, batchState *BatchState, intermediateUsedCounters *vm.Counters) (*vm.BatchCounterCollector, error) {
```
3th param is no longer needed

hexoscott replied:
Will remove it

hexoscott replied:
We still need this at the moment, because if we're closing a partial batch we will have loaded up the counters from the previous run.

kstoykov replied:
Oh, I see. Ok.

kstoykov commented on zk/stages/stage_sequence_execute_data_stream.go:
```diff
@@ -78,29 +83,38 @@ func (sbc *SequencerBatchStreamWriter) writeBlockDetailsToDatastream(verifiedBun
+	return isLastEntryBatchEnd, nil
+}
+
+func finalizeLastBatchInDatastreamIfNotFinalized(batchContext *BatchContext, batchState *BatchState, thisBlock uint64) (bool, error) {
```
This function is longer used. Is this expected?

hexoscott replied:
Yup yup, the same code is basically replicated in the runBatchEnd function so as we're about to do that immediately after we detect we're in a bad state it made sense to cut it out

kstoykov replied:
The "finalizeLastBatchInDatastreamIfNotFinalized" is used when converting a RPC node to Sequencer node. The idea is to finalize the previous batch in the DS. So If we no longer need it, then its logic must be handled by something else.

hexoscott committed:
refactor of checking for stream gap in sequencing

hexoscott committed:
do not connect to datastream during startup

this causes timeout problems on new nodes that could spend a long time running L1 sync

hexoscott committed:
refactor of batch end logic in sequencing

hexoscott committed:
tidy up and comments around new datastream handling in sequencer
